core/LSU.cpp

```diff
@@ -482,16 +905,56 @@ namespace olympia
             }
         }
 
-        sparta_assert(false, "Attempt to complete instruction no longer exiting in issue queue!");
+        sparta_assert(false, "Attempt to complete instruction no longer exiting in issue queue! " << inst_ptr);
+    }
+
+    void inline LSU::appendToQueue_(LoadStoreIssueQueue &queue, const LoadStoreInstInfoPtr &inst_info_ptr){
+        sparta_assert(queue.size() < ldst_inst_queue_size_,
+                      "Appending load queue causes overflows!");
+
+        bool exists = std::find_if(
+            queue.begin(), queue.end(),
+            [inst_info_ptr](auto &queue_inst_ptr) { return queue_inst_ptr->getInstPtr() == inst_info_ptr->getInstPtr(); }
+        ) == std::end(queue);
+
+        sparta_assert(exists,  "Cannot push duplicate instructions into the queue " << inst_info_ptr);
```

Comment:
If you add a variable to the `LoadStoreInstInfo` that captures where it is, you can double check to make sure it's not already enqueued:
```
    sparta_assert(false == inst_info_ptr->isInIssueQueue(), "Instruction already in queue: " << inst_info_ptr);
    inst_info_ptr->setInIssueQueue(true);
```

This removes the expensive `find_if`